src/plugins/discourse/fetch.js

```diff
@@ -0,0 +1,212 @@
+// @flow
+
+/**
+ * Class for retrieving data from the Discourse API.
+ *
+ * The Discourse API implements the JSON endpoints for all functionality of the actual site.
+ * As such, it tends to return a lot of information that we don't care about (in contrast
+ * to a GraphQL API which would give us only what we ask for). As such, we implement a simple
+ * interface over it, which both abstracts over calling the API, and does some post-processing
+ * on the results to simplify it to data that is relevant for us.
+ */
+
+import stringify from "json-stable-stringify";
+import fetch from "isomorphic-fetch";
+
+export type UserId = number;
+export type PostId = number;
+export type TopicId = number;
+
+export type Topic = {|
+  +id: TopicId,
+  +title: string,
+  +timestampMs: number,
+  +authorUsername: string,
+|};
+
+export type Post = {|
+  +id: PostId,
+  +topicId: TopicId,
+  // Which number post this was within the topic (starts at 1)
+  +postNumber: number,
+  // The postNumber of the post within the same topic that this post was a
+  // reply to. Will be `null` if this post was the first post, or if it was a
+  // reply to the first post.
+  +replyToPostNumber: number | null,
+  +timestampMs: number,
+  +authorUsername: string,
+|};
+
+export type TopicWithPosts = {|
+  +topic: Topic,
+  // Not guaranteed to contain all the Posts in the topic—clients will need to
+  // manually request some posts. The raw response actually includes a list of
+  // all the PostIds in the topic, but for now we don't use them.
+  //
+  // We do use these Posts though, as it allows us to save requesting them all
+  // individually.
+  +posts: $ReadOnlyArray<Post>,
+|};
+
+/**
+ * Interface over the external Discourse API, structured to suit our particular needs.
+ * We have an interface (as opposed to just an implementation) to enable easy mocking and
+ * testing.
```

Comment:
Well, and also the Discourse `Mirror` should implement this interface,
right?

Reply:
No, this interface is the right interface if you want to do an update (find out what the latest indices are, compare to stored, and fetch new stuff). If you want to create a graph from scratch, then you just want methods that give you every topic and every post so you can iterate over them conveniently. So the mirror has a different interface.